stan/math/rev/core/vari.hpp

```diff
@@ -300,12 +314,115 @@ class vari_value<T, require_eigen_dense_base_t<T>> : public vari_base {
    */
   inline void init_dependent() { adj_.setOnes(); }
 
+  /**
+   * Implimentation for assignable objects that sets the adjoint values to zero.
+   */
+  template <typename Checker = T,
+            require_t<bool_constant<!std::is_const<Checker>::value>>* = nullptr>
+  inline void set_zero_adjoint_impl() {
+    adj_.setZero();
+  }
+  /**
+   * Implimentation for views which performs a no-op
+   */
+  template <typename Checker = T, require_t<std::is_const<Checker>>* = nullptr>
+  inline void set_zero_adjoint_impl() {}
+
   /**
    * Set the adjoint value of this variable to 0.  This is used to
    * reset adjoints before propagating derivatives again (for
    * example in a Jacobian calculation).
    */
-  inline void set_zero_adjoint() final { adj_.setZero(); }
+  inline void set_zero_adjoint() final { set_zero_adjoint_impl(); }
+
+  /**
+   * A block view of the underlying Eigen matrices.
+   * @param i Starting row of block.
+   * @param j Starting columns of block.
+   * @param p Number of rows to return.
+   * @param q Number of columns to return.
+   */
+  inline const auto block(Eigen::Index i, Eigen::Index j, Eigen::Index p,
+                          Eigen::Index q) const {
+    const auto& val_block = val_.block(i, j, p, q);
+    const auto& adj_block = adj_.block(i, j, p, q);
+    return vari_value<const PlainObject, Eigen::OuterStride<>>(
+        val_block, adj_block, val_block.outerStride(), adj_block.outerStride());
+  }
+
+  /**
+   * View of the head of Eigen vector types.
+   * @param n Number of elements to return from top of vector.
+   */
+  inline const auto head(Eigen::Index n) const {
+    return vari_value<const PlainObject>(val_.head(n), adj_.head(n));
+  }
+
+  /**
+   * View of the tail of the Eigen vector types.
+   * @param n Number of elements to return from bottom of vector.
+   */
+  inline const auto tail(Eigen::Index n) const {
+    return vari_value<const PlainObject>(val_.tail(n), adj_.tail(n));
```

Comment:
Possibly a dumb question (warning).

If I'm understanding the implementation, these views initialise a new ```vari_value``` - which allocates new memory for the values and adjoints to be 'viewed' and copies the values/adjoints into that. Have I got that right?

If so, given that the ```vari_values``` contain pointers to memory for the values/adjoints which is then ```Eigen:Map```-ed, would it be possible to instead create a ```vari_value``` specialisation that is instantiated with pointers to the memory needed? That way the internal ```Eigen::Map``` is directly mapping the memory from the Matrix being 'viewed', rather than allocating new memory and copying into it?

Does that make sense? I haven't entirely grok-ed the static matrix stuff, so I might be off here

Reply:
I think the major bit to note here is the protected constructor for making these views

```cpp
 protected:
  template <typename S, typename K, typename StrideS = Eigen::Stride<0, 0>,
            typename StrideK = Eigen::Stride<0, 0>,
            require_convertible_t<S&, value_type>* = nullptr,
            require_convertible_t<K&, value_type>* = nullptr>
  explicit vari_value(S&& val, K&& adj,
                      StrideS val_stride = Eigen::Stride<0, 0>(),
                      StrideK adj_stride = Eigen::Stride<0, 0>())
      : val_(val.data(), val.rows(), val.cols(), val_stride),
        adj_(adj.data(), adj.rows(), adj.cols(), adj_stride) {}
```

That doesn't copy or allocate any memory. Instead it just constructs a map that is given a copy of the pointer (8 bytes), the dimensions of the array, and the stride (the blocks ops are essentially just maps with a specific stride). Does that make sense? It's a bit weird for sure but I think I nipped this right so we don't make copies or alloc new mem

Reply:
Though I think you are correct that we may want another class just to handle the views

Reply:
Ah missed that part, that would do it! I'll pipe down now